src/decode_words.jl

```diff
@@ -88,44 +86,128 @@ function control_data(TLM_HOW_Data::TLM_HOW_Data_Struct,
-        dc.subframes_decoded[2] = true
-        TLM_HOW, subframe_3_data = decode_subframe_3(words_in_subframes[3])
-        dc.subframes_decoded[3] = true
-        TLM_HOW = decode_subframe_4(words_in_subframes[4])
-        dc.subframes_decoded[4] = true
-        TLM_HOW = decode_subframe_5(words_in_subframes[5])
-        dc.subframes_decoded[5] = true
-
-
-        if control_data(TLM_HOW, subframe_1_data, subframe_2_data, subframe_3_data)  # Checks for Data Errors
-            data = create_data(TLM_HOW, subframe_1_data, subframe_2_data, subframe_3_data) # Creates GPSData from single, subframe specific data and the TLM Word
-            dc.data = data
-            println(" DECODING COMPLETED!")
+        if dc.subframes_decoded_new == [0,0,0,0,0]
+            TLM_HOW, subframe_1_data = decode_subframe_1(words_in_subframe)
+            dc.data_next = GPSData(TLM_HOW, subframe_1_data)
+            control_state = true
+            dc.subframes_decoded_new[1] = control_state
+            dc.data = GPSData(dc.data, TLM_HOW)
+            if PRINT_ON
+                println("DECODED SUB1!")
+            end
+        elseif dc.subframes_decoded_new == [1,0,0,0,0]
+            TLM_HOW, subframe_2_data = decode_subframe_2(words_in_subframe)
+            control_state = control_data_sub2(dc.data_next, subframe_2_data)
+            dc.subframes_decoded_new[2] = control_state
+            if control_state
+                dc.data_next = GPSData(dc.data_next, TLM_HOW, subframe_2_data)
+                dc.data = GPSData(dc.data, TLM_HOW)
+                if PRINT_ON
+                    println("DECODED SUB2!")
+                end
+            end
+        elseif dc.subframes_decoded_new == [1,1,0,0,0]
+            TLM_HOW, subframe_3_data = decode_subframe_3(words_in_subframe)
+            control_state = control_data_sub3(dc.data_next, subframe_3_data)
+            dc.subframes_decoded_new[3] = control_state
+            if control_state
+                dc.data_next = GPSData(dc.data_next, TLM_HOW, subframe_3_data)
+                dc.data = GPSData(dc.data, TLM_HOW)
+                if PRINT_ON
+                    println("DECODED SUB3!")
+                end
+            end
+        elseif dc.subframes_decoded_new == [1,1,1,0,0]
+            TLM_HOW = decode_subframe_4(words_in_subframe)
+            control_state = true
+            dc.subframes_decoded_new[4] = control_state
+            dc.data = GPSData(dc.data, TLM_HOW)
+            if PRINT_ON
+                println("DECODED SUB4!")
+            end
+        elseif dc.subframes_decoded_new == [1,1,1,1,0]
+            TLM_HOW = decode_subframe_5(words_in_subframe)
+            dc.data_next = GPSData(dc.data_next, TLM_HOW)
+            control_state = true
+            dc.subframes_decoded_new[5] = control_state
+            if PRINT_ON
+                println("DECODING OF FRAME FINISHED!")
+            end
+        end
```

Comment:
Instead of waiting for Subframe 1 and then decode Subframe by Subframe, I would suggest decoding TLM and HOW here to detect which subframe is present and then directly decode that subframe. After that, the IODC/IODE have to be compared to verify that the data is from the same set of ephemeris. The `de.subframes_decoded` register should still be used to detect when the decoding is ready. For the ephemeris, this is the case after subframe 1 to 3. After each subframe is decoded once, the decoding should nevertheless be performed in the order of the subframes in order to minimize data losses due to non matching IODC/IODE.